File: sdk/android/src/main/kotlin/com/swifter/fraudshield/binding/DeviceBindingManager.kt
package com.swifter.fraudshield.binding

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

/**
 * DeviceBindingManager
 *
 * Orchestrates device registration and validation against the PayGuard backend.
 * Handles:
 *  - Signal collection + fingerprint generation
 *  - Secure token storage in Android Keystore
 *  - POST /v1/device/register (on first login)
 *  - POST /v1/device/validate (on every login + transaction)
 *  - Step-up authentication enforcement
 */
class DeviceBindingManager(
    private val context: Context,
    private val apiBaseUrl: String,
    private val apiKey: String,
) {
    private val fingerprintCollector = DeviceFingerprintCollector(context)
    private val tokenStore           = KeystoreTokenStore(context)

    data class BindingResult(
        val deviceStatus: String,   // "trusted" | "new_device" | "suspicious_device"
        val requiresStepUp: Boolean,
        val stepUpMethod: String?,  // "otp_verification" | "biometric" | "email_verification"
        val riskDelta: Int,
        val triggeredRules: List<String>,
    )

    /**
     * Register this device for the given user.
     * Must be called after a successful login to bind the device.
     * Stores the returned device token securely in Android Keystore.
     */
    suspend fun registerDevice(userId: String): BindingResult = withContext(Dispatchers.IO) {
        val signals     = fingerprintCollector.collect()
        val fingerprint = fingerprintCollector.generateFingerprint(signals)

        val body = buildRegisterBody(userId, fingerprint, signals)
        val response = postJson("$apiBaseUrl/v1/device/register", body)

        val token  = response.getString("device_token")
        val status = response.getString("device_status")

        // Store token securely in Keystore
        tokenStore.storeToken(token)

        BindingResult(
            deviceStatus   = status,
            requiresStepUp = status != "trusted",
            stepUpMethod   = response.optString("required_action").ifEmpty { null },
            riskDelta      = 0,
            triggeredRules = emptyList(),
        )
    }

    /**
     * Validate this device on every login and transaction.
     * Returns the risk delta to merge into the transaction risk score.
     * If step-up is required, the app should prompt before proceeding.
     */
    suspend fun validateDevice(userId: String): BindingResult = withContext(Dispatchers.IO) {
        val storedToken = tokenStore.retrieveToken()
        val signals     = fingerprintCollector.collect()
        val fingerprint = fingerprintCollector.generateFingerprint(signals)

        if (storedToken == null) {
            // No token stored — register as new device
            return@withContext registerDevice(userId)
        }

        val body = buildValidateBody(userId, storedToken, fingerprint, signals)
        val response = postJson("$apiBaseUrl/v1/device/validate", body)

        val status    = response.getString("device_status")
        val riskDelta = response.optInt("risk_delta", 0)
        val rules     = response.optJSONArray("triggered_rules")
        val rulesList = (0 until (rules?.length() ?: 0)).map { rules!!.getString(it) }

        // If token validation failed, re-register
        if (response.optString("validation_error").isNotEmpty()) {
            tokenStore.clearToken()
            return@withContext registerDevice(userId)
        }

        BindingResult(
            deviceStatus   = status,
            requiresStepUp = status != "trusted",
            stepUpMethod   = response.optString("required_action").ifEmpty { null },
            riskDelta      = riskDelta,
            triggeredRules = rulesList,
        )
    }

    /** Clear stored token on logout. */
    fun clearBinding() = tokenStore.clearToken()

    /** True if a device token is currently stored. */
    fun hasBinding(): Boolean = tokenStore.hasToken()

    // ── HTTP helpers ──────────────────────────────────────────────────────────

    private fun postJson(urlStr: String, body: JSONObject): JSONObject {
        val url  = URL(urlStr)
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            setRequestProperty("Content-Type", "application/json")
            setRequestProperty("x-api-key", apiKey)
            doOutput = true
            connectTimeout = 5_000
            readTimeout    = 8_000
        }
        OutputStreamWriter(conn.outputStream, Charsets.UTF_8).use {
            it.write(body.toString())
        }

        val responseCode = conn.responseCode
        val responseBody = if (responseCode < 400) {
            conn.inputStream.bufferedReader().readText()
        } else {
            conn.errorStream?.bufferedReader()?.readText() ?: "{}"
        }
        conn.disconnect()
        return JSONObject(responseBody)
    }

    private fun buildRegisterBody(
        userId: String,
        fingerprint: String,
        signals: DeviceFingerprintCollector.DeviceSignals,
    ) = JSONObject().apply {
        put("user_id", userId)
        put("signals", JSONObject().apply {
            put("device_model",      signals.deviceModel)
            put("os_version",        signals.osVersion)
            put("screen_resolution", signals.screenResolution)
            put("app_install_id",    signals.appInstallId)
            put("sim_country",       signals.simCountry)
            put("carrier",           signals.carrier)
            put("ip_address",        signals.ipAddress)
            put("timezone",          signals.timezone)
            put("locale",            signals.locale)
            put("is_rooted",         signals.isRooted)
            put("is_emulator",       signals.isEmulator)
            put("is_jailbroken",     signals.isJailbroken)
            put("app_hash",          signals.appHash)
        })
    }

    private fun buildValidateBody(
        userId: String,
        token: String,
        fingerprint: String,
        signals: DeviceFingerprintCollector.DeviceSignals,
    ) = JSONObject().apply {
        put("user_id",            userId)
        put("device_token",       token)
        put("device_fingerprint", fingerprint)
        put("ip_address",         signals.ipAddress)
        put("sim_country",        signals.simCountry)
    }
}

File: sdk/ios/Sources/FraudShieldSDK/Binding/DeviceBindingManager.swift
import Foundation

/// DeviceBindingManager (iOS/Swift)
///
/// Orchestrates device registration and validation against the PayGuard backend.
/// Mirrors the Android DeviceBindingManager API exactly.
public final class DeviceBindingManager {

    private let apiBaseUrl: String
    private let apiKey:     String
    private let collector   = DeviceFingerprintCollector()
    private let tokenStore  = KeychainTokenStore()

    public struct BindingResult {
        public let deviceStatus:    String   // "trusted" | "new_device" | "suspicious_device"
        public let requiresStepUp:  Bool
        public let stepUpMethod:    String?  // "otp_verification" | "biometric" | "email_verification"
        public let riskDelta:       Int
        public let triggeredRules:  [String]
    }

    public init(apiBaseUrl: String, apiKey: String) {
        self.apiBaseUrl = apiBaseUrl
        self.apiKey     = apiKey
    }

    // MARK: - Public API

    /// Register this device for the given user.
    /// Call after successful login. Stores token securely in Keychain.
    public func registerDevice(userId: String) async throws -> BindingResult {
        let signals     = collector.collect()
        let fingerprint = collector.generateFingerprint(signals)

        let body = buildRegisterBody(userId: userId, fingerprint: fingerprint, signals: signals)
        let response = try await postJSON(path: "/v1/device/register", body: body)

        guard let token  = response["device_token"]  as? String,
              let status = response["device_status"]  as? String else {
            throw BindingError.invalidResponse
        }

        try tokenStore.storeToken(token)

        return BindingResult(
            deviceStatus:   status,
            requiresStepUp: status != "trusted",
            stepUpMethod:   response["required_action"] as? String,
            riskDelta:      0,
            triggeredRules: []
        )
    }

    /// Validate this device. Call on every login and before every transaction.
    public func validateDevice(userId: String) async throws -> BindingResult {
        guard let storedToken = tokenStore.retrieveToken() else {
            return try await registerDevice(userId: userId)
        }

        let signals     = collector.collect()
        let fingerprint = collector.generateFingerprint(signals)

        let body = buildValidateBody(
            userId: userId, token: storedToken,
            fingerprint: fingerprint, signals: signals
        )
        let response = try await postJSON(path: "/v1/device/validate", body: body)

        // If validation error, re-register
        if let _ = response["validation_error"] as? String {
            tokenStore.clearToken()
            return try await registerDevice(userId: userId)
        }

        let status    = response["device_status"]  as? String ?? "new_device"
        let riskDelta = response["risk_delta"]      as? Int    ?? 0
        let rules     = response["triggered_rules"] as? [String] ?? []

        return BindingResult(
            deviceStatus:   status,
            requiresStepUp: status != "trusted",
            stepUpMethod:   response["required_action"] as? String,
            riskDelta:      riskDelta,
            triggeredRules: rules
        )
    }

    /// Clear stored token on logout.
    public func clearBinding() { tokenStore.clearToken() }

    /// True if a device token is stored.
    public var hasBinding: Bool { tokenStore.hasToken() }

    // MARK: - Private helpers

    private func postJSON(path: String, body: [String: Any]) async throws -> [String: Any] {
        guard let url = URL(string: "\(apiBaseUrl)\(path)") else {
            throw BindingError.invalidURL
        }

        var request = URLRequest(url: url, timeoutInterval: 8)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, _) = try await URLSession.shared.data(for: request)
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw BindingError.invalidResponse
        }
        return json
    }

    private func buildRegisterBody(userId: String, fingerprint: String,
                                   signals: DeviceSignals) -> [String: Any] {
        return [
            "user_id": userId,
            "signals": [
                "device_model":      signals.deviceModel,
                "os_version":        signals.osVersion,
                "screen_resolution": signals.screenResolution,
                "app_install_id":    signals.appInstallId,
                "sim_country":       signals.simCountry,
                "carrier":           signals.carrier,
                "ip_address":        signals.ipAddress,
                "timezone":          signals.timezone,
                "locale":            signals.locale,
                "is_rooted":         signals.isRooted,
                "is_emulator":       signals.isEmulator,
                "is_jailbroken":     signals.isJailbroken,
                "app_hash":          signals.appHash,
            ]
        ]
    }

    private func buildValidateBody(userId: String, token: String,
                                   fingerprint: String, signals: DeviceSignals) -> [String: Any] {
        return [
            "user_id":            userId,
            "device_token":       token,
            "device_fingerprint": fingerprint,
            "ip_address":         signals.ipAddress,
            "sim_country":        signals.simCountry,
        ]
    }

    public enum BindingError: Error {
        case invalidURL
        case invalidResponse
    }
}
